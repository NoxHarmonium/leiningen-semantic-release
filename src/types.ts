import { either } from "fp-ts/lib/Either";
import { lstatSync } from "fs";
import * as t from "io-ts";
import { withFallback } from "io-ts-types/lib/withFallback";
import { Context } from "semantic-release";

const directoryPathCodec = new t.Type<string, string, unknown>(
  "DirectoryPath",
  t.string.is,
  (u, c) =>
    either.chain(t.string.validate(u, c), s => {
      return lstatSync(s).isDirectory()
        ? t.failure(u, c, "not a path to a valid directory")
        : t.success(s);
    }),
  String
);

export type DirectoryPath = t.TypeOf<typeof directoryPathCodec>;

export const pluginConfigCodec = t.type({
  skipDeploy: withFallback(t.boolean, true),
  pkgRoot: directoryPathCodec,
  uberJar: withFallback(t.boolean, false)
});

export type PluginConfig = t.TypeOf<typeof pluginConfigCodec>;

export type PluginStep = (
  config: PluginConfig,
  context: Context
) => Promise<void>;
