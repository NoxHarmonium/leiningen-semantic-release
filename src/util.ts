import execa from "execa";
import { Context } from "semantic-release";
import * as pkg from "../package.json";

/**
 * Execute while streaming to stdout in realtime
 */
export const exec = async (
  file: string,
  args: readonly string[],
  cwd: string
) => {
  const childProcess = execa(file, args, { cwd });
  if (childProcess.stdout !== null) {
    childProcess.stdout.pipe(process.stdout);
  }
  if (childProcess.stderr !== null) {
    childProcess.stderr.pipe(process.stderr);
  }
  return childProcess.then(({ stdout }) => stdout);
};

export const printVersion = ({ logger }: Context) => {
  logger.log(`Running ${pkg.name} version ${pkg.version}`);
};
