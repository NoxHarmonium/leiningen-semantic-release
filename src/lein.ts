import { Context } from "semantic-release";
import { PluginConfig } from "./types";
import { exec } from "./util";

/**
 * Gets the path to the "lein" executable.
 * Could be extended in the future to try different locations.
 */
const findCommand = async () => "lein"

export const updateVersionInLeinProject = async (
  { pkgRoot }: PluginConfig,
  { logger, nextRelease }: Context
) => {
  if (nextRelease === undefined) {
    throw new Error("context.nextRelease is undefined");
  }
  const version = nextRelease.version;
  logger.log(`Updating project.clj to version ${version}`);
  const command = await findCommand();
  return exec(command, ["change", "version", "set", `"${version}"`], pkgRoot);
};

export const deploy = async (
  { pkgRoot }: PluginConfig,

  { logger, nextRelease }: Context
) => {
  if (nextRelease === undefined) {
    throw new Error("context.nextRelease is undefined");
  }
  const version = nextRelease.version;
  logger.log("Deploying version %s with leiningen", version);
  const command = await findCommand();
  return exec(command, ["deploy"], pkgRoot);
};

export const getVersion = async (
  { pkgRoot }: PluginConfig,
  { logger }: Context
) => {
  logger.log("Sanity checking project version with leiningen");
  const command = await findCommand();
  return exec(
    command,
    // Uses the lein-project-version plugin to print the version number
    // Only updates the project in memory, the project file on disk will remain unaffected
    [
      "update-in",
      ":plugins",
      "conj",
      '[lein-project-version "0.1.0"]',
      "--",
      "project-version"
    ],
    pkgRoot
  );
};

export const leinPackage = async (
  { uberJar, pkgRoot }: PluginConfig,
  { logger }: Context
) => {
  logger.log("Packaging with leiningen");
  const command = await findCommand();
  return exec(command, [uberJar ? "uberjar" : "jar"], pkgRoot);
};
