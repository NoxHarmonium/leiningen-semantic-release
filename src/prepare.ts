import { Context } from "semantic-release";
import { leinPackage, updateVersionInLeinProject } from "./lein";
import { PluginConfig, PluginStep } from "./types";
import { printVersion } from "./util";

export const prepareStep: PluginStep = async (
  config: PluginConfig,
  context: Context
) => {
  printVersion(context);

  // set and commit version number in pom.xml
  await updateVersionInLeinProject(config, context);
  await leinPackage(config, context);
};
