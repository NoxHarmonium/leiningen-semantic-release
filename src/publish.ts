import { Context } from "semantic-release";
import { deploy } from "./lein";
import { PluginConfig, PluginStep } from "./types";
import { printVersion } from "./util";

export const publishStep: PluginStep = async (
  config: PluginConfig,
  context: Context
) => {
  printVersion(context);

  if (!config.skipDeploy) {
    // deploy the project to maven-central
    await deploy(config, context);
  }
};
