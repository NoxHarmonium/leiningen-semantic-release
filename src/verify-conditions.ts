import { Context } from "semantic-release";
import { getVersion } from "./lein";
import { PluginConfig, PluginStep } from "./types";
import { printVersion } from "./util";

/**
 * Verify that the leiningen project is properly setup to allow deployment
 */
export const verifyConditionsStep: PluginStep = async (
  config: PluginConfig,
  context: Context
) => {
  printVersion(context);

  await getVersion(config, context);
};
