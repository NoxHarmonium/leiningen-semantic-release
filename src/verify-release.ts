import { Context } from "semantic-release";
import { valid } from "semver";
import { getVersion } from "./lein";
import { PluginConfig, PluginStep } from "./types";
import { printVersion } from "./util";

export const verifyReleaseStep: PluginStep = async (
  config: PluginConfig,
  context: Context
) => {
  const { logger } = context;
  printVersion(context);

  const projectVersion = await getVersion(config, context);

  // check integrity of pom version
  if (valid(projectVersion) === null) {
    logger.log(
      "WARNING: project.clj version of %s is an invalid semver version",
      projectVersion
    );
  }
};
