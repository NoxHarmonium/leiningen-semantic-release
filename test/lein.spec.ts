import { Context, NextRelease } from "semantic-release";
import { PluginConfig } from "../src/types";
import mockContext from "./__fixtures__/mock-context";

// tslint:disable: no-duplicate-string

const mockUtilModule = {
  exec: jest.fn(),
  printVersion: jest.fn()
};
jest.mock("../src/util", () => mockUtilModule);

import {
  deploy,
  getVersion,
  leinPackage,
  updateVersionInLeinProject
} from "../src/lein";

type LeinWrapperFunction = (
  config: PluginConfig,
  context: Context
) => Promise<string>;

const performTestPass = async (
  fn: LeinWrapperFunction,
  uberJar: boolean,
  nextRelease: NextRelease
) => {
  await fn(
    {
      pkgRoot: "pkgroot",
      skipDeploy: false,
      uberJar
    },
    {
      ...mockContext,
      nextRelease
    }
  );
  expect(mockUtilModule.exec).toBeCalledTimes(1);
  expect(mockUtilModule.exec.mock.calls[0]).toMatchSnapshot();
};

const performTestFail = async (fn: LeinWrapperFunction) => {
  expect.assertions(1);
  try {
    await fn(
      {
        pkgRoot: "pkgroot",
        skipDeploy: false,
        uberJar: true
      },
      mockContext
    );
  } catch (e) {
    expect(e).toBeInstanceOf(Error);
  }
};

describe("when calling leiningen", () => {
  beforeEach(() => {
    mockUtilModule.exec.mockReset();
    mockUtilModule.printVersion.mockReset();
  });
  describe("when calling lein wrapper methods", () => {
    describe("when nextRelease is defined", () => {
      test.each([
        ["updateVersionInLeinProject", updateVersionInLeinProject, false],
        ["updateVersionInLeinProject", updateVersionInLeinProject, true],
        ["deploy", deploy, false],
        ["deploy", deploy, true],
        ["getVersion", getVersion, false],
        ["getVersion", getVersion, true],
        ["leinPackage", leinPackage, false],
        ["leinPackage", leinPackage, true]
      ])(
        "'%s' should call lein with the correct arguments",
        // @ts-ignore
        async (_: string, fn: LeinWrapperFunction, uberJar: boolean) => {
          await performTestPass(fn, uberJar, {
            version: "1.2.3",
            gitTag: "v1.2.3-git",
            gitHead: "abcd1234",
            notes: "some notes"
          });
        }
      );
    });
    describe("when nextRelease is undefined", () => {
      test.each([
        ["updateVersionInLeinProject", updateVersionInLeinProject],
        ["deploy", deploy]
      ])(
        "'%s' should throw",
        // @ts-ignore
        async (_: string, fn: LeinWrapperFunction) => {
          await performTestFail(fn);
        }
      );
    });
  });
});
