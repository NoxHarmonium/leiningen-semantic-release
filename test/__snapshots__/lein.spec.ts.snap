// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when calling leiningen when calling lein wrapper methods when nextRelease is defined 'deploy' should call lein with the correct arguments 1`] = `
Array [
  "lein",
  Array [
    "deploy",
  ],
  "pkgroot",
]
`;

exports[`when calling leiningen when calling lein wrapper methods when nextRelease is defined 'deploy' should call lein with the correct arguments 2`] = `
Array [
  "lein",
  Array [
    "deploy",
  ],
  "pkgroot",
]
`;

exports[`when calling leiningen when calling lein wrapper methods when nextRelease is defined 'getVersion' should call lein with the correct arguments 1`] = `
Array [
  "lein",
  Array [
    "update-in",
    ":plugins",
    "conj",
    "[lein-project-version \\"0.1.0\\"]",
    "--",
    "project-version",
  ],
  "pkgroot",
]
`;

exports[`when calling leiningen when calling lein wrapper methods when nextRelease is defined 'getVersion' should call lein with the correct arguments 2`] = `
Array [
  "lein",
  Array [
    "update-in",
    ":plugins",
    "conj",
    "[lein-project-version \\"0.1.0\\"]",
    "--",
    "project-version",
  ],
  "pkgroot",
]
`;

exports[`when calling leiningen when calling lein wrapper methods when nextRelease is defined 'leinPackage' should call lein with the correct arguments 1`] = `
Array [
  "lein",
  Array [
    "jar",
  ],
  "pkgroot",
]
`;

exports[`when calling leiningen when calling lein wrapper methods when nextRelease is defined 'leinPackage' should call lein with the correct arguments 2`] = `
Array [
  "lein",
  Array [
    "uberjar",
  ],
  "pkgroot",
]
`;

exports[`when calling leiningen when calling lein wrapper methods when nextRelease is defined 'updateVersionInLeinProject' should call lein with the correct arguments 1`] = `
Array [
  "lein",
  Array [
    "change",
    "version",
    "set",
    "\\"1.2.3\\"",
  ],
  "pkgroot",
]
`;

exports[`when calling leiningen when calling lein wrapper methods when nextRelease is defined 'updateVersionInLeinProject' should call lein with the correct arguments 2`] = `
Array [
  "lein",
  Array [
    "change",
    "version",
    "set",
    "\\"1.2.3\\"",
  ],
  "pkgroot",
]
`;
